### PR TITLE
Add timestamp rules and constraints to decoder in Whisper example

### PR DESCRIPTION
**Pull request by rsb-tbg**
### Summary

This PR implements proper timestamp handling rules in the Whisper decoder to ensure accurate and well-formed timestamp generation during transcription, according to: https://github.com/openai/whisper/blob/e8622f9afc4eba139bf796c210f5c01081000472/whisper/decoding.py#L439

### New Example Output

```
cargo run --example whisper --release --features="symphonia,metal"
    Finished `release` profile [optimized] target(s) in 0.44s
     Running `target/release/examples/whisper`
No audio file submitted: Downloading https://huggingface.co/datasets/Narsil/candle_demo/blob/main/samples_jfk.wav
pcm data loaded 176000
loaded mel: [1, 80, 3000]
0.0s -- 30.0s
  0.0s-1.8s: 
  1.8s-10.4s:  and so my fellow Americans ask not what your country can do for you ask what you can do for your country
```

### Changes Made

**Core Timestamp Rules Implementation**

- Added apply_timestamp_rules() method that enforces OpenAI Whisper's timestamp constraints
- Implemented 4 key timestamp rules:
     1. Timestamp Pairing: Timestamps must come in pairs, except directly before EOT
     2. Non-decreasing Order: Timestamps cannot decrease in value
     3. Forced Initial Timestamp: Suppress non-timestamp tokens at the beginning of transcription
     4. Probability-based Preference: When timestamp probability sum exceeds any individual token, only consider timestamps

### Configuration Enhancements
- Added max_initial_timestamp_index parameter to limit the maximum initial timestamp value
- Changed timestamps default to true (was previously disabled by default)
- Updated decoder constructor to accept the new timestamp constraint parameter

### Code Structure Improvements
- Refactored decoder methods to use self.model consistently instead of mutable borrows
- Enhanced timestamp token detection and validation logic
- Added comprehensive masking system for applying multiple timestamp constraints

### Benefits
✅ More accurate timestamp generation following OpenAI Whisper specifications
✅ Prevents malformed timestamp sequences
✅ Configurable initial timestamp constraints for better control
✅ Improved transcription quality with proper temporal alignment

### Testing
The implementation follows the timestamp rules from the original OpenAI Whisper codebase, ensuring compatibility and correctness.

### CODE REVIEW

File: candle-examples/examples/whisper/main.rs [hunk 1]
```diff
@@ -265,6 +273,161 @@ impl Decoder {
...
+                .iter()
+                .filter(|&&t| t >= timestamp_begin)
+                .cloned()
+                .collect();
+
+            if !timestamp_tokens.is_empty() {
+                let timestamp_last = if last_was_timestamp && !penultimate_was_timestamp {
+                    *timestamp_tokens.last().unwrap()
+                } else {
+                    timestamp_tokens.last().unwrap() + 1
```
[the remark this answers is not in the document]
Comment (rsb-tbg) [reply candle-examples/examples/whisper/main.rs:346]: This logic is correct because:

When last_was_timestamp && !penultimate_was_timestamp = we just output a timestamp and need to force a text token next
In this case, we use the exact timestamp value (no +1) to prevent generating the same timestamp again
In all other cases, we add +1 to prevent going backwards and ensure forward progress
This matches the Python implementation exactly. Copilot's suggestion would break the timestamp pairing rules.

File: candle-examples/examples/whisper/main.rs [hunk 2]
```diff
@@ -265,6 +273,161 @@ impl Decoder {
...
+            logits = logits.broadcast_add(&mask)?;
+        }
+
+        // ========== RULE 4: Probability-based timestamp preference ==========
+        // If sum of probability over timestamps is above any other token, sample timestamp
+        let log_probs = softmax(&logits, 0)?;
+        let log_probs_vec: Vec<f32> = log_probs.to_vec1()?;
+
+        // Sum probabilities for timestamp tokens
+        let timestamp_prob_sum: f32 = log_probs_vec[timestamp_begin as usize..].iter().sum();
```
[the remark this answers is not in the document]
Comment (rsb-tbg) [reply candle-examples/examples/whisper/main.rs]: resolved in subsequent commit

File: candle-examples/examples/whisper/main.rs [hunk 3]
```diff
@@ -265,6 +273,161 @@ impl Decoder {
...
+
+        // Find max probability for non-timestamp tokens
+        let max_text_prob = log_probs_vec[..timestamp_begin as usize]
```
[the remark this answers is not in the document]
Comment (rsb-tbg) [reply candle-examples/examples/whisper/main.rs]: resolved in subsequent comment via log_softmax

**Comment by greenrazer** [comment]
Thanks!
